### PR TITLE
[SUSTAIN-632] Do not update the users table to contain the public key. The keys table should be the only source of truth

### PR DESCRIPTION
**Pull request by PrajaktaPurohit**
TODO:
 - [x] Add sqitch tag
 - [x] Add omnitruck changes
 - [x] If `knife create client` is the only way to create clients, this bug does not affect clients.
 - [x] Update function returns [[<<"update_user">>, <<>>]] that does not get handled in our code
```
==> /var/log/opscode/opscode-erchef/current <==
2017-08-23_22:48:46.28882 [error] {<<"method=PUT; path=/users/test; status=500; ">>,{error,{error,{case_clause,[[{<<"update_user">>,<<>>}]]},[{chef_db,update,3,[{file,"/var/cache/omnibus/src/oc_erchef/_build/default/lib/chef_db/src/chef_db.erl"},{line,259}]},{oc_chef_wm_base,update_from_json,4,[{file,"/var/cache/omnibus/src/oc_erchef/_build/default/lib/oc_chef_wm/src/oc_chef_wm_base.erl"},{line,1121}]},{webmachine_resource,resource_call,3,[{file,"/var/cache/omnibus/src/oc_erchef/_build/default/lib/webmachine/src/webmachine_resource.erl"},{line,203}]},{webmachine_resource,do,3,[{file,"/var/cache/omnibus/src/oc_erchef/_build/default/lib/webmachine/src/webmachine_resource.erl"},{line,148}]},{webmachine_decision_core,resource_call,1,[{file,"/var/cache/omnibus/src/oc_erchef/_build/default/lib/webmachine/src/webmachine_decision_core.erl"},{line,47}]},{webmachine_decision_core,accept_helper,1,[{file,"/var/cache/omnibus/src/oc_erchef/_build/default/lib/webmachine/src/webmachine_decision_core.erl"},{line,613}]},{webmachine_decision_core,decision,1,[{file,"/var/cache/omnibus/src/oc_erchef/_build/default/lib/webmachine/src/webmachine_decision_core.erl"},{line,518}]},{webmachine_decision_core,handle_request,2,[{file,"/var/cache/omnibus/src/oc_erchef/_build/default/lib/webmachine/src/webmachine_decision_core.erl"},{line,32}]}]}}}
```
- [x] Add user function returns [[<<"add_user">>,<<>>]] that does not get handled in sqerl
```
(oc_erchef@api)1> [error] {<<"method=POST; path=/users; status=500; ">>,{error,{error,{case_clause,{ok,[[{<<"add_user">>,<<>>}]]}},[{sqerl,statement,4,[{file,"/host/src/oc_erchef/_build/default/lib/sqerl/src/sqerl.erl"},{line,111}]},{timer,tc,1,[{file,"timer.erl"},{line,166}]},{stats_hero,ctime,3,[{file,"/host/src/oc_erchef/_build/default/lib/stats_hero/src/stats_hero.erl"},{line,131}]},{chef_db,create,3,[{file,"/host/src/oc_erchef/_build/dev/lib/chef_db/src/chef_db.erl"},{line,168}]},{oc_chef_wm_base,create_from_json,5,[{file,"/host/src/oc_erchef/_build/dev/lib/oc_chef_wm/src/oc_chef_wm_base.erl"},{line,1061}]},{webmachine_resource,resource_call,3,[{file,"/host/src/oc_erchef/_build/default/lib/webmachine/src/webmachine_resource.erl"},{line,203}]},{webmachine_resource,do,3,[{file,"/host/src/oc_erchef/_build/default/lib/webmachine/src/webmachine_resource.erl"},{line,148}]},{webmachine_decision_core,resource_call,1,[{file,"/host/src/oc_erchef/_build/default/lib/webmachine/src/webmachine_decision_core.erl"},{line,47}]}]}}}
```

http://wilson.ci.chef.co/job/chef-server-trigger-ad_hoc/10/downstreambuildview/

### CODE REVIEW

File: src/oc_erchef/apps/chef_db/src/chef_sql.erl [hunk 1]
```diff
@@ -1004,7 +1004,7 @@ create_object(insert_cookbook_artifact_version = QueryName, Args) when is_list(A
         {conflict, _Why} = Conflict -> Conflict
     end;
 create_object(QueryName, Args) when is_atom(QueryName), is_list(Args) ->
-    sqerl:statement(QueryName, Args, count);
+    sqerl:select(QueryName, Args, count);
```
Comment (marcparadise) [inline src/oc_erchef/apps/chef_db/src/chef_sql.erl]: Why did you need to change this to `select`?  Context for concern: creation of most other object types is done via insert statements, and this would affect them as well.

File: src/oc_erchef/schema/sqitch.plan [hunk 2]
```diff
@@ -80,3 +80,7 @@ node_policy_name_policy_group 2015-09-04T01:15:03Z Daniel DeLeo <ddeleo@lorentz.
 
 users_email_functional_index [users] 2017-05-30T12:10:32Z Stephan Renatus <srenatus@chef.io> # Add functional index for index-based lower(email) lookups
 @users_email_functional_index 2017-05-30T13:27:15Z Stephan Renatus <srenatus@chef.io> # Add functional index for index-based lower(email) lookups
+
+create_and_update_users 2017-08-23T22:53:21Z Prajakta Purohit <prajakta@chef.io> # Adding functions to create and update users by inserting a sentinel value in the public_key_columns
+keys_update_trigger [keys_update_trigger@users_email_functional_index] 2017-08-23T23:12:29Z Prajakta Purohit <prajakta@chef.io> # The insert and update triggers ignore rows with sentinel value for public_key.
```
Comment (marcparadise) [inline src/oc_erchef/schema/sqitch.plan]: The creation of this change as `@users_email_functional_index` seems off to me since we want it to go in sequence after the previous one - and if sqitch tracks this as dependent on users_email_functional_index, it may not go in that order.  Could you verify that this won't be a thing?

File: src/oc_erchef/apps/chef_db/src/chef_db.erl [hunk 3]
```diff
@@ -167,6 +166,7 @@ create(ObjectRec0, #context{server_api_version = ApiVersion, reqid = ReqId}, Act
     Fields = chef_object:fields_for_insert(ObjectRec2),
     case stats_hero:ctime(ReqId, {chef_sql, create_object},
                           fun() -> chef_sql:create_object(QueryName, Fields) end) of
+        {ok,[[{<<"add_user">>,<<>>}]]} -> ok;
```
Comment (marcparadise) [inline src/oc_erchef/apps/chef_db/src/chef_db.erl]: This is very specific to the add_user query/function.  If we are not able to get a return value consistent with other inserts (number of rows inserted), it's worth a closer look at epgsql+sqerl to understand the  returned `[[{binary(), binary()}]]` from stored proc calls. Would it be sufficient to match similar to the below w/ guards:

File: src/oc_erchef/schema/deploy/create_and_update_users.sql [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+                                       p_external_authentication_uid users.external_authentication_uid%TYPE,
+                                       p_recovery_authentication_enabled users.recovery_authentication_enabled%TYPE,
+                                       p_email users.email%TYPE,
+                                       p_username users.username%TYPE,
+                                       p_last_updated_by users.last_updated_by%TYPE,
+                                       p_updated_at users.updated_at%TYPE,
+                                       p_admin users.admin%TYPE,
+                                       p_id character(32)) RETURNS integer AS $update_user$
+    BEGIN
+    IF p_public_key != 'thisisnotakey' THEN
```
[the remark this answers is not in the document]
Comment (PrajaktaPurohit) [reply src/oc_erchef/schema/deploy/create_and_update_users.sql]: The reason we are not using null is - null in v0 api implies that the user wants to delete the key, hence the use of a sentinel value.

File: src/oc_erchef/apps/chef_db/src/chef_sql.erl (hunk 1 above)
Comment (PrajaktaPurohit) [reply src/oc_erchef/apps/chef_db/src/chef_sql.erl]: We changed the insert statements to be a function in plpgsql, and those return in the form : {ok,[[{<<"add_user">>,<<>>}]]}
but select expects return in the form:  https://github.com/chef/sqerl/blob/master/src/sqerl.erl#L94-L101
which is why the change from statement to select.

File: src/oc_erchef/schema/sqitch.plan (hunk 2 above)
Comment (PrajaktaPurohit) [reply src/oc_erchef/schema/sqitch.plan]: That is a direct change due to `sqitch rework keys_update_trigger` but will make sure.

File: src/oc_erchef/schema/revert/keys_update_trigger.sql [hunk 5]
```diff
@@ -1,14 +1,45 @@
...
+     IF NEW.public_key IS NOT NULL THEN
+       INSERT INTO keys
+         (id, key_name, public_key, key_version, created_at,  expires_at) VALUES
+         (NEW.id, 'default', NEW.public_key, NEW.pubkey_version, now(), 'infinity'::timestamp);
+     END IF;
+     RETURN NULL; -- result is ignored since this is an AFTER trigger
+   END;
+$add_key$ LANGUAGE plpgsql;
+
+CREATE OR REPLACE FUNCTION update_key() RETURNS TRIGGER AS $update_key$
```
[the remark this answers is not in the document]
Comment (marcparadise) [reply src/oc_erchef/schema/revert/keys_update_trigger.sql]: Disregard, it was misnamed from the start and changing it now will complicate the sqitch bits unnecessarily.

File: src/oc_erchef/schema/revert/keys_update_trigger@users_email_functional_index.sql [hunk 6]
```diff
@@ -0,0 +1,14 @@
+-- Revert keys_update_trigger
```
[the remark this answers is not in the document]
Comment (PrajaktaPurohit) [reply src/oc_erchef/schema/revert/keys_update_trigger@users_email_functional_index.sql]: right the current keys_update_trigger.sql does that, this is the revert for the file before making my changes.

File: src/oc_erchef/apps/chef_db/src/chef_sql.erl (hunk 1 above)
Comment (PrajaktaPurohit) [reply src/oc_erchef/apps/chef_db/src/chef_sql.erl]: updated as discussed!

File: src/oc_erchef/schema/deploy/create_and_update_users.sql [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+                        key_version = p_pubkey_version,
+                        expires_at = 'infinity'::timestamp
+        WHERE id = p_id AND key_name = 'default';
+       INSERT INTO keys (id, key_name, public_key, key_version, created_at, expires_at)
+         SELECT p_id, 'default', p_public_key, p_pubkey_version, now(), 'infinity'::timestamp
+         WHERE NOT EXISTS (SELECT 1 FROM keys WHERE id = p_id AND key_name = 'default');
+      ELSE
+       DELETE FROM keys WHERE id = p_id AND key_name = 'default';
+      END IF;
+    END IF;
```
[the remark this answers is not in the document]
Comment (PrajaktaPurohit) [reply src/oc_erchef/schema/deploy/create_and_update_users.sql]: we still need this because we will not be inserting the key into the users table in all the new requests (after this release) so just leaving the trigger in will update 'thisisnotakey' into the keys table. This function does the right things and the trigger will only be useful for older versions of chef-server

[the remark this answers is not in the document]
Comment (markan) [reply src/oc_erchef/schema/deploy/create_and_update_users.sql]: In SQL NULL returns NULL for all comparisons. So the DELETE will never fire above I think, but instead NULL != to 'thisisnotakey' will return NULL, which is falsy, and the inner if will never be reached.

```
CREATE OR REPLACE FUNCTION test(val text)
RETURNS text
AS $$
DECLARE
  result text;
BEGIN
   IF val != 'A' THEN
     IF val IS NOT NULL THEN
      result := 'NOT A, NOT NULL';
     ELSE 
      result := 'NOT A, BUT NULL';
     END IF;
   ELSE
      result := 'A';
   END IF;
   RETURN result;
END;
$$;
```

```
opscode_chef=# select test(NULL);
 test 
------
 A
```

[the remark this answers is not in the document]
Comment (markan) [reply src/oc_erchef/schema/deploy/create_and_update_users.sql]: We probably should have unit test coverage of some of these functions to insure they work as expected.

File: src/oc_erchef/schema/deploy/keys_update_trigger.sql [hunk 8]
```diff
@@ -5,29 +5,35 @@
...
-   BEGIN
-     IF NEW.public_key IS NOT NULL THEN
-       INSERT INTO keys
-         (id, key_name, public_key, key_version, created_at,  expires_at) VALUES
-         (NEW.id, 'default', NEW.public_key, NEW.pubkey_version, now(), 'infinity'::timestamp);
-     END IF;
-     RETURN NULL; -- result is ignored since this is an AFTER trigger
-   END;
+  BEGIN
+    IF NEW.public_key != 'thisisnotakey' THEN
```
Comment (markan) [inline src/oc_erchef/schema/deploy/keys_update_trigger.sql]: See the NULL comparision comment for create_and_update_users.sql above

File: src/oc_erchef/schema/deploy/keys_update_trigger.sql [hunk 9]
```diff
@@ -5,29 +5,35 @@
...
-         SELECT NEW.id, 'default', NEW.public_key, NEW.pubkey_version, now(), 'infinity'::timestamp
-         WHERE NOT EXISTS (SELECT 1 FROM keys WHERE id = NEW.id AND key_name = 'default');
-     ELSE
-       DELETE FROM keys WHERE id = NEW.id AND key_name = 'default';
-     END IF;
-     RETURN NULL; -- result is ignored since this is an AFTER trigger
-   END;
+  BEGIN
+    IF NEW.public_key != 'thisisnotakey' THEN
+      IF NEW.public_key IS NOT NULL THEN
```
Comment (markan) [inline src/oc_erchef/schema/deploy/keys_update_trigger.sql]: See the NULL comparision comment for create_and_update_users.sql above

File: src/oc_erchef/apps/chef_db/priv/pgsql_statements.config [hunk 10]
```diff
@@ -484,17 +484,10 @@
 %% User-related
 %%
 {insert_user,
-  <<"INSERT INTO users (id, authz_id, username, email, hashed_password, salt,
-     hash_type, last_updated_by, created_at, updated_at, external_authentication_uid,
-      recovery_authentication_enabled, serialized_object, admin, pubkey_version) VALUES
-      ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, false, 0)">>}.
+  <<"SELECT add_user($1, $2, $3, $4, 'thisisnotakey', 0, $5, $6, $7, $8, $9, $10, $11, $12, $13, false)">>}.
```
[the remark this answers is not in the document]
Comment (PrajaktaPurohit) [reply src/oc_erchef/apps/chef_db/priv/pgsql_statements.config]: done!

File: src/oc_erchef/schema/deploy/create_and_update_users.sql (hunk 7 above)
[the remark this answers is not in the document]
Comment (PrajaktaPurohit) [reply src/oc_erchef/schema/deploy/create_and_update_users.sql]: That is an awesome catch! updated the cases to fall through to the delete operation when expected.

I will check the unit tests in bifrost and add something to test this functionality.

File: src/oc_erchef/schema/deploy/keys_update_trigger.sql (hunk 8 above)
Comment (PrajaktaPurohit) [reply src/oc_erchef/schema/deploy/keys_update_trigger.sql]: updated.

File: src/oc_erchef/apps/chef_db/src/chef_db.erl [hunk 11]
```diff
@@ -168,6 +167,7 @@ create(ObjectRec0, #context{server_api_version = ApiVersion, reqid = ReqId}, Act
     case stats_hero:ctime(ReqId, {chef_sql, create_object},
                           fun() -> chef_sql:create_object(QueryName, Fields) end) of
         {ok, 1} -> ok;
+        {ok, 0} -> <<"Record not created '", ObjectRec2/binary, "'.">>;
```
Comment (stevendanna) [inline src/oc_erchef/apps/chef_db/src/chef_db.erl]: The dialyzer spec on this function needs to be updated.  Perhaps it would be good to use `{error, <<"Record..._`  here as you did below. I think this case would be pretty rare as I imagine if we can't create what we are trying to create, we would get an error back rather than a successful query?

File: src/oc_erchef/schema/deploy/create_and_update_users.sql [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+      DELETE FROM keys WHERE id = p_id AND key_name = 'default';
+    END IF;
+    IF p_public_key != 'this_is_not_a_key' THEN
+      UPDATE keys SET public_key = p_public_key,
+                      key_version = p_pubkey_version,
+                      expires_at = 'infinity'::timestamp
+      WHERE id = p_id AND key_name = 'default';
+        INSERT INTO keys (id, key_name, public_key, key_version, created_at, expires_at)
+        SELECT p_id, 'default', p_public_key, p_pubkey_version, now(), 'infinity'::timestamp
+        WHERE NOT EXISTS (SELECT 1 FROM keys WHERE id = p_id AND key_name = 'default');
```
Comment (stevendanna) [inline src/oc_erchef/schema/deploy/create_and_update_users.sql]: I know this is how we are doing "upsert" in the original code, but I think this construction is racy.  My guess is that concurrent user updates are rare enough that we can stay with this though.

File: src/oc_erchef/schema/deploy/create_and_update_users.sql [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+        (id, authz_id, username, email, public_key, hashed_password, salt, hash_type,
+         last_updated_by, created_at, updated_at, external_authentication_uid,
+         recovery_authentication_enabled, serialized_object, admin,
+         pubkey_version)
+       VALUES (p_id, p_authz_id, p_username, p_email, 'this_is_not_a_key',
+         p_hashed_password, p_salt, p_hash_type, p_last_updated_by, p_created_at, p_updated_at,
+         p_external_authentication_uid, p_recovery_authentication_enabled,
+         p_serialized_object, p_admin, p_pubkey_version) RETURNING 1)
+      SELECT count(*) FROM createduser INTO inserteduser;
+      RETURN inserteduser;
```
Comment (christophermaier) [inline src/oc_erchef/schema/deploy/create_and_update_users.sql]: You don't need this indirection of using CTEs to always insert the literal value `1` into a variable.

Why does this function need to return anything if it's always returning `1`? Is the desire to indicate whether something was inserted?

File: src/oc_erchef/schema/revert/create_and_update_users.sql [hunk 14]
```diff
@@ -0,0 +1,8 @@
+-- Revert enterprise_chef:create_and_update_users from pg
+
+BEGIN;
+
+DROP FUNCTION IF EXISTS add_user();
+DROP FUNCTION IF EXISTS update_user();
```
Comment (christophermaier) [inline src/oc_erchef/schema/revert/create_and_update_users.sql]: The deploy script implies that these functions existed in the system before, but this revert will not replace those older implementations.

File: src/oc_erchef/schema/revert/keys_update_trigger.sql [hunk 15]
```diff
@@ -1,14 +1,45 @@
...
+   BEGIN
+     IF NEW.public_key IS NOT NULL THEN
+       UPDATE keys SET public_key = NEW.public_key, key_version = NEW.pubkey_version, expires_at = 'infinity'::timestamp
+         WHERE id = NEW.id AND key_name = 'default';
+       INSERT INTO  keys (id, key_name, public_key, key_version, created_at, expires_at)
+         SELECT NEW.id, 'default', NEW.public_key, NEW.pubkey_version, now(), 'infinity'::timestamp
+         WHERE NOT EXISTS (SELECT 1 FROM keys WHERE id = NEW.id AND key_name = 'default');
```
Comment (christophermaier) [inline src/oc_erchef/schema/revert/keys_update_trigger.sql]: Looks like you're trying to do an upsert? Can you use the `ON CONFLICT` clause instead (see https://www.postgresql.org/docs/current/static/sql-insert.html)... not sure what version of PG y'all are on these days).

I saw this was being done above, as well.

File: src/oc_erchef/schema/t/test_create_and_update_users.sql [hunk 16]
```diff
@@ -0,0 +1,216 @@
+CREATE OR REPLACE FUNCTION test_zzzzcreate_and_update_users()
```
Comment (christophermaier) [inline src/oc_erchef/schema/t/test_create_and_update_users.sql]: Why "zzzz"?

File: src/oc_erchef/schema/t/test_create_and_update_users.sql [hunk 17]
```diff
@@ -0,0 +1,216 @@
...
+RETURNS SETOF TEXT
+LANGUAGE plpgsql
+AS $$
+BEGIN
+
+-- v0 add_user tests --
+RETURN QUERY SELECT is(add_user('11111111111111111111111111111111',
+                                '22222222222222222222222222222222',
+                                'aaaa',
```
Comment (christophermaier) [inline src/oc_erchef/schema/t/test_create_and_update_users.sql]: For these dummy values, I'd recommend having them describe what they are as much as you can given any constraints that exist on the data.

So if "aaaa" is a username, use "test_user" instead, etc.

File: src/oc_erchef/schema/t/test_create_and_update_users.sql [hunk 18]
```diff
@@ -0,0 +1,216 @@
...
+    ARRAY['this_is_not_a_key'],
+    'unchanged public_key in users table for v0'
+);
+
+PREPARE "after update email v0" AS SELECT email from users
+                                     WHERE id='11111111111111111111111111111111';
+RETURN QUERY SELECT results_eq(
+    '"after update email v0"',
+    ARRAY['changed_email_address_v0@chef.io'],
+    'after update email address for v0'
```
Comment (christophermaier) [inline src/oc_erchef/schema/t/test_create_and_update_users.sql]: I'd recommend stating these messages in terms of the desired behavior you're going after, instead of the mechanics of what is happening. That makes it easier to figure out what the overall intention of the code is.

File: src/oc_erchef/schema/deploy/create_and_update_users.sql (hunk 12 above)
Comment (christophermaier) [reply src/oc_erchef/schema/deploy/create_and_update_users.sql]: Ah, I commented on this down below... if it's available on the Postgres version you're using, an `ON CONFLICT` clause would be better for this.

Alternatively, using CTEs for this might be preferable, as you could contain the operations within a single statement.

You're also in PL/pgSQL, so you could also use exception handling, but that's more expensive.

File: src/oc_erchef/schema/deploy/create_and_update_users.sql [hunk 19]
```diff
@@ -0,0 +1,105 @@
...
+      VALUES
+         (p_id, 'default', p_public_key, p_pubkey_version, now(), 'infinity'::timestamp);
+    END IF;
+    WITH createduser AS
+      (INSERT INTO users
+        (id, authz_id, username, email, public_key, hashed_password, salt, hash_type,
+         last_updated_by, created_at, updated_at, external_authentication_uid,
+         recovery_authentication_enabled, serialized_object, admin,
+         pubkey_version)
+       VALUES (p_id, p_authz_id, p_username, p_email, 'this_is_not_a_key',
```
Comment (christophermaier) [inline src/oc_erchef/schema/deploy/create_and_update_users.sql]: If the intention is to have the `keys` table be the source of truth for keys, why keep key-related information in the `users` table at all? Is this step 1 of a broader refactoring?

File: src/oc_erchef/schema/t/test_create_and_update_users.sql (hunk 16 above)
Comment (PrajaktaPurohit) [reply src/oc_erchef/schema/t/test_create_and_update_users.sql]: wanted that test to run last - will edit.

File: src/oc_erchef/schema/deploy/create_and_update_users.sql (hunk 19 above)
Comment (PrajaktaPurohit) [reply src/oc_erchef/schema/deploy/create_and_update_users.sql]: yep - sort of a phase1 since on env like hosted we need to support older versions of chef-server.

File: src/oc_erchef/schema/deploy/create_and_update_users.sql (hunk 13 above)
Comment (PrajaktaPurohit) [reply src/oc_erchef/schema/deploy/create_and_update_users.sql]: yea - to make sure that the insert did happen.

File: src/oc_erchef/schema/deploy/create_and_update_users.sql (hunk 12 above)
Comment (stevendanna) [reply src/oc_erchef/schema/deploy/create_and_update_users.sql]: Unfortunately, I'm not sure we can drop support for 9.2 yet (we should check with product) which means no `ON CONFLICT`.  I think overall using `CTE`s would also make some of this a bit more readable; Perhaps we can do that as a more easily reviewable follow-up improvement. (Getting a bit worried that there is a lot going on here as this is already code not many know well.)

File: src/oc_erchef/schema/t/test_create_and_update_users.sql (hunk 18 above)
Comment (PrajaktaPurohit) [reply src/oc_erchef/schema/t/test_create_and_update_users.sql]: updated!

File: src/oc_erchef/schema/t/test_create_and_update_users.sql (hunk 17 above)
Comment (PrajaktaPurohit) [reply src/oc_erchef/schema/t/test_create_and_update_users.sql]: updated!

File: src/oc_erchef/schema/deploy/create_and_update_users.sql (hunk 12 above)
Comment (PrajaktaPurohit) [reply src/oc_erchef/schema/deploy/create_and_update_users.sql]: I have added the new syntax as a comment in there - a TODO for the future!

File: src/oc_erchef/schema/deploy/create_and_update_users.sql (hunk 13 above)
Comment (PrajaktaPurohit) [reply src/oc_erchef/schema/deploy/create_and_update_users.sql]: RE: You don't need this indirection of using CTEs to always insert the literal value 1 into a variable. - will update that but maybe as a second pass, if that is ok?

File: src/oc_erchef/schema/revert/create_and_update_users.sql (hunk 14 above)
Comment (PrajaktaPurohit) [reply src/oc_erchef/schema/revert/create_and_update_users.sql]: Yes - this is just safeguarding for the future, we don't expect the functions to actually be present.

File: src/oc_erchef/schema/revert/keys_update_trigger.sql (hunk 15 above)
Comment (PrajaktaPurohit) [reply src/oc_erchef/schema/revert/keys_update_trigger.sql]: added a comment can be changes once we drop support for postgres 9.2/9.3

File: src/oc_erchef/schema/t/test_create_and_update_users.sql (hunk 17 above)
Comment (PrajaktaPurohit) [reply src/oc_erchef/schema/t/test_create_and_update_users.sql]: done!

File: src/oc_erchef/schema/t/test_create_and_update_users.sql (hunk 18 above)
Comment (PrajaktaPurohit) [reply src/oc_erchef/schema/t/test_create_and_update_users.sql]: done!

File: src/oc_erchef/schema/revert/keys_update_trigger.sql (hunk 15 above)
Comment (PrajaktaPurohit) [reply src/oc_erchef/schema/revert/keys_update_trigger.sql]: Actually maybe I will leave this one as is - since it is the revert of the update trigger.

File: src/oc_erchef/apps/chef_db/src/chef_db.erl [hunk 20]
```diff
@@ -258,6 +258,7 @@ update(#chef_cookbook_version{org_id =OrgId} = Record, #context{reqid = ReqId} =
 update(ObjectRec, #context{reqid = ReqId}, ActorId) ->
     case stats_hero:ctime(ReqId, {chef_sql, do_update},
                           fun() -> chef_sql:update(ObjectRec, ActorId) end) of
+        0 -> {error, <<"Record not updated '", ObjectRec/binary, "'.">>};
```
Comment (marcparadise) [inline src/oc_erchef/apps/chef_db/src/chef_db.erl]: It looks like the changes [do_update](https://github.com/chef/chef-server/pull/1383/files#diff-43908ac1f0b634347760bb3d12a9cba8R1100) will return `{error, Reason}` - can we remove this branch from here? 

Same question for `create`, looks like the {ok. 0} return is handled there.

File: src/oc_erchef/apps/chef_db/src/chef_sql.erl [hunk 21]
```diff
@@ -1081,6 +1092,16 @@ delete_object(delete_cookbook_by_orgid_name = Query, OrgId, Name) ->
 update(ObjectRec, ActorId) ->
     chef_object:update(ObjectRec, ActorId, fun select_rows/1).
 
+do_update(QueryName, UpdateFields) when QueryName =:= update_user_by_id;
+                                        QueryName =:= update_user_by_id_v0,
+                                        is_list(UpdateFields)  ->
+    case sqerl:select(QueryName, UpdateFields, first_as_scalar, [update_user]) of
+        {ok, 1} -> {ok, 1};
+        {ok, 0} -> {error, <<"Record not updated '", (term_to_binary(QueryName))/binary, "'.">>};
+        {ok, none} -> {ok, not_found};
```
Comment (marcparadise) [inline src/oc_erchef/apps/chef_db/src/chef_sql.erl]: the caller seems is expecting `not_found` instead of `{ok, not_found}`.

File: src/oc_erchef/schema/deploy/keys_update_trigger.sql [hunk 22]
```diff
@@ -1,33 +1,43 @@
 -- Deploy keys_update_trigger
 --
 -- This keeps the keys table consistent when the clients and user tables update their keys
---
+-- We need these triggers to maintain compatibility with older versions ok chef-server.
```
Comment (marcparadise) [inline src/oc_erchef/schema/deploy/keys_update_trigger.sql]: s/ok/of

File: src/oc_erchef/schema/deploy/keys_update_trigger@users_email_functional_index.sql [hunk 23]
```diff
@@ -0,0 +1,45 @@
...
+
+CREATE OR REPLACE FUNCTION update_key() RETURNS TRIGGER AS $update_key$
+   BEGIN
+     IF NEW.public_key IS NOT NULL THEN
+       UPDATE keys SET public_key = NEW.public_key, key_version = NEW.pubkey_version, expires_at = 'infinity'::timestamp
+         WHERE id = NEW.id AND key_name = 'default';
+       INSERT INTO  keys (id, key_name, public_key, key_version, created_at, expires_at)
+         SELECT NEW.id, 'default', NEW.public_key, NEW.pubkey_version, now(), 'infinity'::timestamp
+         WHERE NOT EXISTS (SELECT 1 FROM keys WHERE id = NEW.id AND key_name = 'default');
+     ELSE
```
Comment (marcparadise) [inline src/oc_erchef/schema/deploy/keys_update_trigger@users_email_functional_index.sql]: This is a port of the existing logic --  but it at always makes me say, "What what?" when I read it.  Could you add a comment explaining why a null key causes a deletion? (v0 api behavior compatibility)

File: src/oc_erchef/apps/chef_db/src/chef_db.erl (hunk 20 above)
Comment (PrajaktaPurohit) [reply src/oc_erchef/apps/chef_db/src/chef_db.erl]: there are 2 different code paths that call into either the do_update or update functions, so we need this in both places.

Comment (PrajaktaPurohit) [reply src/oc_erchef/apps/chef_db/src/chef_db.erl]: yes!

File: src/oc_erchef/schema/deploy/keys_update_trigger@users_email_functional_index.sql (hunk 23 above)
Comment (PrajaktaPurohit) [reply src/oc_erchef/schema/deploy/keys_update_trigger@users_email_functional_index.sql]: done!
